### PR TITLE
Set g.co/agent in Stackdriver exported via Node info

### PR DESCRIPTION
**Pull request by draffensperger**
This uses the `Language`, `CoreLibraryVersion` and `ExporterVersion` fields of the `LibraryInfo` message on `Node` to set the `g.co/agent` span label in the Stackdriver exporter.

Because this code is running in the OpenCensus service (agent/collector) it's safe to assume that the exporter type in the application library code must have been the `ocagent-exporter`. Hence, it sets the `g.co/agent` string as follows: `opencensus-[Language] [CoreLibraryVersion]; ocagent-exporter [ExporterVersion]`, e.g. `opencensus-python 0.0.1; ocagent-exporter 0.0.2`

Fixes https://github.com/census-instrumentation/opencensus-service/issues/525.

### CODE REVIEW

**Comment by draffensperger** [comment]
@songy23 and @pjanotti would you be open to merging this PR even though the OpenCensus service is being related by the OpenTelemetry service?

I noticed that the OpenTelemetry service doesn't have a Stackdriver exporter yet, so I think merging this PR would make it easier to adopt this feature in OpenTelemetry if we base the Stackdriver exporter on the OpenCensus service one.

**Comment by pjanotti** [comment]
Yes, this should be done first here (OC) then we can pick it up on the contrib side of OpenTelemetry when that gets ready.

File: exporter/stackdriverexporter/stackdriver.go [hunk 1]
```diff
@@ -132,3 +177,36 @@ func (sde *stackdriverExporter) ConsumeMetricsData(ctx context.Context, md data.
...
 	return nil
 }
+
+func agentForLibraryInfo(li *commonpb.LibraryInfo) string {
+	// The exporter must be the OpenCensus agent exporter because the spans
+	// have been written to the OpenCensus service.
+	return "opencensus-" + getLanguageName(li.Language) + " " + li.CoreLibraryVersion + "; ocagent-exporter " + li.ExporterVersion
+}
+
+func getLanguageName(l commonpb.LibraryInfo_Language) string {
```
Comment (pjanotti) [inline exporter/stackdriverexporter/stackdriver.go]: iirc proto can transform this in a string, any specific reason to have this function?

File: exporter/stackdriverexporter/stackdriverexporter_test.go [hunk 2]
```diff
@@ -14,4 +14,124 @@
...
+	"reflect"
+	"testing"
+	"time"
+
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/census-instrumentation/opencensus-service/data"
```
Comment (pjanotti) [inline exporter/stackdriverexporter/stackdriverexporter_test.go]: nit: blank line to separate the import block. Also the repo itself should be the last "block", so the 3 last ones need to be moved to the 2nd block.

File: exporter/stackdriverexporter/stackdriver.go [hunk 3]
```diff
@@ -111,6 +141,21 @@ func StackdriverTraceExportersFromViper(v *viper.Viper) (tps []consumer.TraceCon
 	return
 }
 
+// ExportSpans is the method that translates OpenCensus-Proto Traces into AWS X-Ray spans.
+// It uniquely maintains
+func (sde *stackdriverExporter) pushTraceData(ctx context.Context, td data.TraceData) (int, error) {
+	agent := agentForLibraryInfo(td.Node.GetLibraryInfo())
```
Comment (pjanotti) [inline exporter/stackdriverexporter/stackdriver.go]: in principle LibraryInfo can be nil. While Node shouldn't be nil there was some cases in the past of that so for safety it seems reasonable to consider this possibility.

**Comment by pjanotti** [comment]
It seems that code coverage numbers are wrong and don't reflect that actual PR numbers...

**Comment by draffensperger** [comment]
Thanks for the reviews! I'll may not get to this for several days but I plan to fix it up per the comments.

File: exporter/exporterwrapper/exporterwrapper.go [hunk 4]
```diff
@@ -34,6 +34,12 @@ import (
 	spandatatranslator "github.com/census-instrumentation/opencensus-service/translator/trace/spandata"
 )
 
+// OCSpanExporter is an interface for the ExportSpan function of trace.Exporter.
```
[the remark this answers is not in the document]
Comment (draffensperger) [reply exporter/exporterwrapper/exporterwrapper.go]: Well, I'm open to a better approach, but this was what I seemed to need to do to be able to pass in a fake span exporter for a test. What would you suggest?

File: exporter/stackdriverexporter/stackdriver.go [hunk 5]
```diff
@@ -24,28 +24,52 @@ import (
 	"github.com/spf13/viper"
 	"go.opencensus.io/trace"
 
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
 	"github.com/census-instrumentation/opencensus-service/consumer"
```
[the remark this answers is not in the document]
Comment (draffensperger) [reply exporter/stackdriverexporter/stackdriver.go]: Added

File: exporter/stackdriverexporter/stackdriver.go (hunk 3 above)
Comment (draffensperger) [reply exporter/stackdriverexporter/stackdriver.go]: Thanks for pointing this out. I added checks for `nil` here and similarly below. I also added more test cases to cover the various `nil` values that are possible.

File: exporter/stackdriverexporter/stackdriver.go [hunk 6]
```diff
@@ -111,6 +141,21 @@ func StackdriverTraceExportersFromViper(v *viper.Viper) (tps []consumer.TraceCon
...
+// It uniquely maintains
+func (sde *stackdriverExporter) pushTraceData(ctx context.Context, td data.TraceData) (int, error) {
+	agent := agentForLibraryInfo(td.Node.GetLibraryInfo())
+	agentVal := &tracepb.AttributeValue{
+		Value: &tracepb.AttributeValue_StringValue{
+			StringValue: &tracepb.TruncatableString{Value: agent},
+		},
+	}
+	for _, span := range td.Spans {
+		span.GetAttributes().GetAttributeMap()[agentLabel] = agentVal
```
[the remark this answers is not in the document]
Comment (draffensperger) [reply exporter/stackdriverexporter/stackdriver.go]: Fixed

File: exporter/stackdriverexporter/stackdriver.go (hunk 1 above)
Comment (draffensperger) [reply exporter/stackdriverexporter/stackdriver.go]: Good point, I changed to calling `.String()` on the language enum. Initially I was hoping to make these agent labels match what is set from the libraries by default (e.g. `opencensus-node` rather than `opencensus-node_js`), but I think standardizing on the proto enum names is fine. I did lowercase it so that it's a little more similar to the way that agent labels get set on the per-language exporters.

File: exporter/stackdriverexporter/stackdriverexporter_test.go (hunk 2 above)
Comment (draffensperger) [reply exporter/stackdriverexporter/stackdriverexporter_test.go]: Fixed

**Comment by draffensperger** [comment]
@pjanotti I think I've addressed all the comments (though wasn't sure how to easily remove the `OCSpanExporter` interface). Could you take another look?

File: exporter/exporterwrapper/exporterwrapper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pjanotti) [reply exporter/exporterwrapper/exporterwrapper.go]: I expected that you could do the same with https://github.com/census-instrumentation/opencensus-go/blob/b4a14686f0a98096416fe1b4cb848e384fb2b22b/trace/export.go#L30-L33 since it is exact the same interface... perhaps I'm missing something...

[the remark this answers is not in the document]
Comment (draffensperger) [reply exporter/exporterwrapper/exporterwrapper.go]: Well, as far as I can tell, the `opencensus-go` package is not currently a build dependency of `opencensus-service`. Would it be worth adding it for importing that interface?

[the remark this answers is not in the document]
Comment (draffensperger) [reply exporter/exporterwrapper/exporterwrapper.go]: @pjanotti friendly ping on this. Are you OK with the new interface to avoid the `opencensus-go` import? If so, could you merge this PR?

**Comment by draffensperger** [comment]
@pjanotti friendly ping on this - are you comfortable with keeping the new interface instead of adding a dependency on `opencensus-go`? Would this be ready to merge?

**Comment by pjanotti** [comment]
hi @draffensperger - sorry, missed your reply earlier. I will merge this and cut a release later this week.

**Comment by draffensperger** [comment]
@pjanotti awesome, thanks so much for your help with this!
